toml/src/main/java/com/fasterxml/jackson/dataformat/toml/Parser.java

```diff
@@ -0,0 +1,407 @@
+                 * seem to be covered by the specification.
+                 */
+                TomlArrayNode array = (TomlArrayNode) existing;
+                if (array.closed) {
+                    throw errorContext.atPosition(lexer).generic("Array already closed");
+                }
+                // Only arrays declared by array tables are not closed, and those are always arrays of objects.
+                node = (TomlObjectNode) array.get(array.size() - 1);
+            } else {
+                throw errorContext.atPosition(lexer).generic("Path into existing non-object value of type " + node.getNodeType());
+            }
+        }
+    }
+
+    private JsonNode parseValue(int nextState) throws IOException {
+        TomlToken firstToken = peek();
+        switch (firstToken) {
+            case STRING:
+                String text = lexer.stringBuilder.toString();
+                pollExpected(TomlToken.STRING, nextState);
+                return factory.textNode(text);
+            case TRUE:
+                pollExpected(TomlToken.TRUE, nextState);
+                return factory.booleanNode(true);
+            case FALSE:
+                pollExpected(TomlToken.FALSE, nextState);
+                return factory.booleanNode(false);
+            case OFFSET_DATE_TIME:
+            case LOCAL_DATE_TIME:
+            case LOCAL_DATE:
+            case LOCAL_TIME:
+                return parseDateTime(nextState);
+            case FLOAT:
+                return parseFloat(nextState);
+            case INTEGER:
+                return parseInt(nextState);
+            case ARRAY_OPEN:
+                return parseArray(nextState);
+            case INLINE_TABLE_OPEN:
+                return parseInlineTable(nextState);
+            default:
+                throw errorContext.atPosition(lexer).unexpectedToken(firstToken, "value");
+        }
+    }
+
+    private JsonNode parseDateTime(int nextState) throws IOException {
+        String text = lexer.yytext();
+        TomlToken token = poll(nextState);
+        // the time-delim index can be [Tt ]. java.time supports only [Tt]
+        if ((token == TomlToken.LOCAL_DATE_TIME || token == TomlToken.OFFSET_DATE_TIME) && text.charAt(10) == ' ') {
+            text = text.substring(0, 10) + 'T' + text.substring(11);
+        }
+
+        if (options.parseTemporalAsJavaTime) {
+            Temporal value;
+            if (token == TomlToken.LOCAL_DATE) {
+                value = LocalDate.parse(text);
+            } else if (token == TomlToken.LOCAL_TIME) {
+                value = LocalTime.parse(text);
+            } else {
+                if (token == TomlToken.LOCAL_DATE_TIME) {
+                    value = LocalDateTime.parse(text);
+                } else if (token == TomlToken.OFFSET_DATE_TIME) {
+                    value = OffsetDateTime.parse(text);
+                } else {
+                    throw new AssertionError();
```

Comment:
I'd prefer converting all assert/AssertionErrors into different kind of internal exception, just something that is not `Error` (since I think catching `Error` is usually no-no, compared to `Exception`).
Jackson-core has `VersionUtil.throwInternal()` which throws a RuntimeException, for example, but need not use that one. Probably best to first add a helper method in this class, call that, let it throw whatever is deemed best.

Reply:
the code path should never be executed, so throwInternal sounds good.